File: src/sokoenginepy/common/sokoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, List, Optional, Tuple

from .characters import Characters
from .config import Config
from .direction import Direction
from .tessellation import index_1d, index_column, index_row, is_on_board_2d
from .tessellation_impl import PusherStepData, TessellationImpl


class SokobanTessellation(TessellationImpl):
    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, PusherStepData]] = {
        Characters.l: PusherStepData(Direction.LEFT, False),
        Characters.L: PusherStepData(Direction.LEFT, True),
        Characters.r: PusherStepData(Direction.RIGHT, False),
        Characters.R: PusherStepData(Direction.RIGHT, True),
        Characters.u: PusherStepData(Direction.UP, False),
        Characters.U: PusherStepData(Direction.UP, True),
        Characters.d: PusherStepData(Direction.DOWN, False),
        Characters.D: PusherStepData(Direction.DOWN, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[PusherStepData, str]] = {
        v: k for k, v in _CHR_TO_PUSHER_STEP.items()
    }

    _NEIGHBOR_SHIFT: Final[List[Tuple[Optional[int], Optional[int]]]] = [
        (-1, 0),  # UP
        (None, None),
        (0, 1),  # RIGHT
        (None, None),
        (1, 0),  # DOWN
        (None, None),
        (0, -1),  # LEFT
        (None, None),
    ]

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT[direction.value]

        if row_shift is None or column_shift is None:
            raise ValueError(
                f"Unsupported direction {direction} for {self.__class__.__name__}"
            )

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS

File: src/libsokoengine/common/sokoban_tessellation.cpp
/// @file
#include "sokoban_tessellation.hpp"

#include "characters.hpp"
#include "pusher_step.hpp"

using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace implementation {

static const directions_t SOK_LEGAL_DIRECTIONS = {
  Direction::LEFT, Direction::RIGHT, Direction::UP, Direction::DOWN};

const directions_t &SokobanTessellation::legal_directions() const {
  return SOK_LEGAL_DIRECTIONS;
}

position_t SokobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  position_t row = index_y(position, width), column = index_x(position, width);
  switch (direction) {
    case Direction::LEFT:
      column += -1;
      break;
    case Direction::UP:
      row += -1;
      break;
    case Direction::RIGHT:
      column += 1;
      break;
    case Direction::DOWN:
      row += 1;
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + to_str(direction) + " for SokobanTessellation!"
      );
  }

  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

char SokobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? Characters::L : Characters::l;
    case Direction::UP:
      return rv.is_push_or_pull() ? Characters::U : Characters::u;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? Characters::R : Characters::r;
    case Direction::DOWN:
      return rv.is_push_or_pull() ? Characters::D : Characters::d;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + to_str(rv.direction())
        + " in SokobanTessellation!"

      );
  }
}

PusherStep SokobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case Characters::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case Characters::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);
    case Characters::u:
      return PusherStep(Direction::UP, Config::NO_ID);
    case Characters::U:
      return PusherStep(Direction::UP, Config::DEFAULT_ID);
    case Characters::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case Characters::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);
    case Characters::d:
      return PusherStep(Direction::DOWN, Config::NO_ID);
    case Characters::D:
      return PusherStep(Direction::DOWN, Config::DEFAULT_ID);
    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in SokobanTessellation!"
      );
  }
}

} // namespace implementation
} // namespace sokoengine
